crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ruma::{MilliSecondsSinceUnixEpoch, OwnedUserId};
+use serde::{Deserialize, Serialize};
+use vodozemac::Ed25519PublicKey;
+
+use crate::types::DeviceKeys;
+
+/// Information on the device and user that sent the megolm session data to us
+///
+/// Sessions start off in `UnknownDevice` state, and progress into `DeviceInfo`
+/// state when we get the device info. Finally, if we can look up the sender
+/// using the device info, the session can be moved into `SenderKnown` state.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum SenderData {
+    /// We have not yet found the (signed) device info for the sending device
+    UnknownDevice {
```

Comment:
It's a bit hard to know if these fields are what we want without seeing how they will be used, hopefully you got it right and won't have to fiddle with migrations :sweat_smile:

Reply:
Yes, I know and maybe I should have done this all together with https://github.com/matrix-org/matrix-rust-sdk/pull/3590 but it just seemed too big.

I did wait until I had https://github.com/matrix-org/matrix-rust-sdk/pull/3590 working before I set this ready to review, so we have at least some chance that I got it right.